### DISCUSSION TITLE
New libcharon plugin that provides the generic kernel 'ipsec' interface for use by User space data plane processes.

### DISCUSSION THREAD

**Discussion by sumkaur** [strongswan/strongswan#1936] [Ideas]
Hello,

Currently there is no interface from libcharon to any user space 'data plane' process (like DPDK, for e.g.). 

libcharon's kernel interface is being implemented by below plugins :-
kernel-netlink - to populate IPSEC info to linux kernel  
kernel-pfkey - to populate IPSEC info to FreeBSD kernel
kernel-wfp - to populate IPSEC info on Windows platform

Can we develop a new libcharon plugin, named - 'kernel-ipsec-dp' for e.g. [ dp stands for 'data-plane ], which can provide a generic IPSEC interface towards a user space data plane process ?

Communication to the user space process will be via some IPC mechanism (like unix socket).
SA, SP information gets sent over this IPC to the user space process. The SA, SP data structures will be aptly and clearly defined.
The user space process listening to this socket reads the SA, SP information and acts accordingly. For e.g. , it may need to modify this info to the way it requires and then use it.

Kindly review this proposal from our side, and share your comments and opinions on this.

Thanks.
Sumit

**Comment by tobiasbrunner** [comment]
> Currently there is no interface from libcharon to any user space 'data plane' process (like DPDK, for e.g.).

There are at least two plugins for VPP (one is currently not public as I've not had time to review it yet, the other is [in the VPP repository](https://github.com/FDio/vpp/tree/master/extras/strongswan/vpp_sswan)).

**Reply by sumkaur** [reply]
OK,

I checked VPP-SSWAN plugin which is part of VPP repository. This plugin is talking to VPP over VPP defined IPSec data structures.

What about a 'generic' IPSEC interface  plugin, for user space dataplane. 
Do you think, this would be possible or useful ?


Thanks
Sumit
